prql-compiler/src/ast/pl/ident.rs

```diff
@@ -153,6 +155,21 @@ pub fn display_ident_part(f: &mut std::fmt::Formatter, s: &str) -> Result<(), st
     }
 }
 
+pub static VALID_IDENT: Lazy<Regex> = Lazy::new(|| {
+    // One of:
+    // - `*`
+    // - An ident starting with `a-z_\$` and containing other characters `a-z0-9_\$`
+    //
+    // We could replace this with pomsky (regex<>pomsky : sql<>prql)
+    // ^ ('*' | [ascii_lower '_$'] [ascii_lower ascii_digit '_$']* ) $
+    Regex::new(r"^((\*)|(^[a-z_\$][a-z0-9_\$]*))$").unwrap()
+});
```

Comment:
I'm confused. Are we using the same regex for PRQL idents and SQL idents? It's quite unlikely that the two have the same lexing rules.

Reply:
This is not a concern of this PR, but just something we should investigate.

Reply:
Although this PR is essentially saying this regex is dealing with the PRQL ident.

Reply:
Yes ... to be frank I was also a bit confused by this. I agree that we should investigate it.

The main reason I moved this regex into the `ident` module is that the `Display` impl of `Ident` indirectly depends on this regex, and I want to move `Ident` to a new `prql_ast` crate in #2994. Currently much of the codebase depends on this Display impl ... I'm planning on removing the dependencies on the Display impl in the future, which will let us move this regex elsewhere.

Reply:
Yes I agree that codegen obviously shouldn't be a dependency of the AST ... however this is only temporarily until I fix that ... because I also believe that errors shouldn't be stringly typed ... the `Reason` enum currently just contains a bunch of strings, which I want to change so that it contains the actual values ... then the formatting happens via a `Display` impl on the Reason type rather than everywhere where the errors are constructed ... and because of this I'll be touching all the places where errors are constructed anyway, and I'd rather only have to touch them once (removing the format string and specifying a proper enum variant) rather than having to touch them twice (firstly using some explicit function call to avoid the Display impl for Expr only to then remove it in the followup change). Does that make sense to you?

tl;dr: Yeah but I want to fix it in a followup PR^^

Reply:
I see, this makes sense.

---

Before you start working on a major refactor of the errors, I'd want us to share the vision of what they are supposed to do.

prql-compiler and PRQL-the-language are work-in-progress, which means that requirements, architecture and some code snippets change quite quickly. Because of that, I would really want that it is easy to construct new errors. By easy, I mean that it is not much harder than just panicking.

The string-contents of the `Error` are quite a nice fit for that, and a text message is exactly what's needed to display the error in an IDE or terminal. Obviously, this is very different if you wanted to integrate the compiler and display the errors with some alternative formatting/UI. But currently, this is not the priority of the development.

I don't want to discourage you - I'm just making sure that my work will not more tedious overnight due to premature optimization. If you think that the refactor can keep the errors easy to construct, you have my blessing to go ahead.

---

PS: We have something that aids "programmatic consumption of errors": error codes, which are used by the vscode extension to match instead of relying on the message. This way, code is part of the "stable public API" and the message is not (i.e. we reserve the freedom to change it anytime).

Reply:
Thanks, ... yes of course that makes sense ... after these PRs are resolved I'll open a draft PR that just converts a few error variants to showcase what I am thinking of :)

Reply:
I have now dropped these regex related changes from this PR since if we don't go with the generic AST approach then we don't have the probem with the Display impl of Expr in the ast crate requiring the regex.